### PR TITLE
chore: add generic metrics support to harness

### PR DESCRIPTION
**Pull request by hamidzr**
## Description

add core api and python sdk support to read and write arbitrary metric types

## Commentary (optional)

depends on https://github.com/determined-ai/determined/pull/7065

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket

DET-9505

### CODE REVIEW

File: harness/determined/common/experimental/determined.py
Comment (azhou-determined) [inline harness/determined/common/experimental/determined.py:383]: is this meant to replace `stream_trials_training_metrics` and `stream_trials_validation_metrics`? if so, should we mark those as deprecated?

File: harness/determined/common/experimental/trial.py
Comment (azhou-determined) [inline harness/determined/common/experimental/trial.py:353]: should this type be just a string? my impression is that `util.MetricType` is for legacy key names? do we even need a separate enum type for it if this is the case?

File: harness/determined/core/_train.py
Comment (azhou-determined) [inline harness/determined/core/_train.py:74]: why was `steps_completed` renamed to `total_batches`?

File: harness/determined/core/_train.py
Comment (azhou-determined) [inline harness/determined/core/_train.py:75]: i'm not a fan of the separate `Metrics` class, makes it harder for users to tell the type.

File: harness/determined/core/_train.py
Comment (azhou-determined) [inline harness/determined/core/_train.py:110]: should this be deprecated?

File: harness/determined/core/_train.py
Comment (azhou-determined) [inline harness/determined/core/_train.py:73]: nit: what about naming this something more generic like `series` or just `metric_name`? `type` implies to me that there's an explicit limited set of "types" allowed

File: harness/determined/core/_train.py
Comment (hamidzr) [reply harness/determined/core/_train.py:74]: we have removed traces of `step`s in our system and there is still open work to remove the last remaining mentions. this is a move toward that.

File: harness/determined/common/experimental/determined.py
Comment (hamidzr) [reply harness/determined/common/experimental/determined.py:383]: do we have an official/preferred way of deprecating symbols in our python codebase? I recently added support for this to the cli. but that's designed for cli use. if not I can add something

File: harness/determined/core/_train.py
Comment (hamidzr) [reply harness/determined/core/_train.py:110]: they don't have to be deprecated but we could deprecate them. I updated it.

File: harness/determined/common/experimental/trial.py
Comment (hamidzr) [reply harness/determined/common/experimental/trial.py:353]: hmm it is not an enum. it's a hint and just an alias to str under the hood IMO this makes it more clear what the argument is about and lets us track the usage of MetricType throughout our codebase by looking at the references of it. you should still be able to pass it raw `str` with no issues.

File: harness/determined/core/_train.py
Comment (hamidzr) [reply harness/determined/core/_train.py:75]: this is more subjective IMO:
some of the same benefits as https://github.com/determined-ai/determined/pull/7117#discussion_r1231462983
but also IDEs are good these days at showing what these are
Screenshot 2023-06-15 at 12 45 50

I could see us wanting something more structured later or or refactor "metrics" which this would help

the type definition could also carry documentation with it in terms of comments, I think we do not allow absolutely anything that fits `Dict[str, Any]` do we?

File: harness/determined/core/_train.py
Comment (hamidzr) [reply harness/determined/core/_train.py:73]: metric_name I think could get confused with what you would nest insdie a metric package eg `loss`
serires could work too. I went with metric_type becuase it maps directly to the apis but we could always deviate if something is more helpful to the user

**Comment by hamidzr** [comment]
I'll do some testing and try to avoid adding new features or support so that I can hand it off to mlsys earlier. how does that sound?

**Comment by hamidzr** [comment]
I think this is ready for a rereview. one question i have is would we like to at some point remove dedicated code and interfaces in our harness for reporting trial and validation metrics? if so we could start warning people as we do here and mark them as deprecated and if not we probably shouldn't start logging warnings

File: harness/determined/common/experimental/determined.py
Comment (azhou-determined) [reply harness/determined/common/experimental/determined.py:383]: i'd just add a log warning:
```
warnings.warn(
    "stream_trials_training_metrics is deprecated, please use stream_trials_metrics instead.",
    FutureWarning,
    stacklevel=2,
)


```

File: harness/determined/common/util.py
Comment (azhou-determined) [inline harness/determined/common/util.py:237]: i'm hesitant about introducing a new method to mark deprecation that isn't used by everything else. 

can we just do what is currently done and log a warning? i see that we also have `common.declarative_argparse.warn_on_usage`. if you do decide to keep the decorator, maybe we can combine these two utils that basically do the same thing?

File: harness/determined/core/_train.py
Comment (azhou-determined) [inline harness/determined/core/_train.py:10]: per the [google style guide](https://google.github.io/styleguide/pyguide.html#22-imports), please import packages, not individual classes/variables

File: harness/determined/common/experimental/trial.py
Comment (azhou-determined) [inline harness/determined/common/experimental/trial.py:79]: why are these getter/setters for steps_completed needed?

File: harness/determined/common/experimental/trial.py
Comment (azhou-determined) [reply harness/determined/common/experimental/trial.py:374]: ah, i thought it was an enum type, my bad. your reasoning about making it easier for us to track in the codebase makes sense, but i'm still hesitant about making it a user-facing type alias. my preference is still for the simple "str" type because it makes it explicit to users what inputs to expect. 

having a type alias here also makes me think there are a limited discrete set of types we support (when any generic string is valid) and i have to go look into the code to figure out what `MetricType` is.

File: harness/determined/common/util.py
Comment (azhou-determined) [inline harness/determined/common/util.py:36]: should these be private? `_LEGACY_TRAINING`?

Comment (azhou-determined) [inline harness/determined/common/util.py:36]: also, these might belong more in `harness.determined.util`?

File: harness/determined/core/_train.py
Comment (azhou-determined) [reply harness/determined/core/_train.py:75]: we allow anything that's json-serializable in metrics. 

yeah, it's pretty subjective. if you prefer the type aliases i won't stop you. my preference is still for explicit types when the object isn't a defined class. i can see how making it a class would be beneficial if later on we refactor it, but i'm also not sure that we _should_ refactor it to be anything more than a `Dict` type since adding complexity to `Metrics` would mean the user would have to construct it. 

looking at other more mature metrics APIs like ([WandB](https://docs.wandb.ai/ref/python/log) and [Ray](https://docs.ray.io/en/latest/ray-air/api/doc/ray.air.session.report.html), it seems like `Dict`/`Dict[str, Any]` is pretty common.

File: harness/determined/core/_train.py
Comment (azhou-determined) [reply harness/determined/core/_train.py]: completely subjective, but `type` just makes me think "what types are supported/exist?" when really here the user is just assigning a label of sorts to a metric. 

"series" seems like a good fit to me, but i'm not bent on it. keeping `metric_type` consistent with the underlying REST APIs makes sense too.

**Comment by azhou-determined** [comment]
since we're deprecating `core.train.report_training_metrics` and `core.train.report_validation_metrics`, should the Trial-side changes for reporting metrics go with this PR too?

File: harness/determined/common/experimental/trial.py
Comment (rb-determined-ai) [reply harness/determined/common/experimental/trial.py:374]: @hamidzr invited me to comment on this.

I am not opposed to all type aliases, but IMO aliasing `str` like this is not very helpful to end users.  If I were a user, I'd start thinking, "I want to call `stream_metrics`" and I'd look at the docs and say, "what the heck is a util.MetricType" and I'd be annoyed if I dug far enough to find out it's just a `str`.

> makes it more clear what the argument is about

I think for parameters especially (since they are inputs provided by the user), the name is what should make the parameter's purpose clear, and the type hint should make it clear what type is expected.  Everybody knows what a `str` is, nobody knows what a `MetricType` is.

Regarding the name, I'd prefer we name this parameter `series` or something.  "Training" isn't a different "type" of metric, it's a different "series" of metrics.  A user could validly think to themselves, "well I want float-type metrics" and be confused as to how to configure `metric_type` accordingly.

> lets us track the usage of MetricType throughout our codebase

That presumes that people consistently use MetricType, even though mypy won't enforce that to be true.  I sympathize with this desire, but I think canonical names for a variable are the best solution for it.  If everything you might assign a `MetricType` to is named `metric_type`, and you can grep for that name.  We're pretty good about keeping canonical names in the harness, I think, such as with symbols like `self.context`, `self.distributed`, or `core_context`.

File: harness/determined/core/_train.py
Comment (rb-determined-ai) [reply harness/determined/core/_train.py]: have the rest apis landed to main yet?  I think we should change the name from `metric_type` to `series`, because I think that name is _way_ better for the concept here.

**Comment by hamidzr** [comment]
> since we're deprecating `core.train.report_training_metrics` and `core.train.report_validation_metrics`, should the Trial-side changes for reporting metrics go with this PR too?

what trial side changes are you thinking about? I'd like to keep the scope smaller so I can leave the rest of the changes up to the team. If it helps I can un-deprecate these.

File: harness/determined/common/experimental/determined.py
Comment (hamidzr) [reply harness/determined/common/experimental/determined.py:383]: soon after my comment I added a `deprecated` decorator. I'm using that now which does log a warning with `category=DeprecationWarning`

File: harness/determined/common/experimental/trial.py
Comment (hamidzr) [reply harness/determined/common/experimental/trial.py:79]: for compatibility with old code using these two since we're moving toward total_batches. Myabe I should have had two copies of these on Training and ValidationMetrics instead.

**Comment by azhou-determined** [comment]
> what trial side changes are you thinking about? I'd like to keep the scope smaller so I can leave the rest of the changes up to the team. If it helps I can un-deprecate these.

all existing calls to `core.train.report_training/validation_metrics` in Trial training loops should be changed to `report_metrics` generic if we're deprecating them. yeah, you can just un-deprecate them and maybe make a ticket to track?

File: harness/determined/core/_train.py
Comment (hamidzr) [reply harness/determined/core/_train.py]: yes, they have but I haven't published the release notes yet.
series does sound good 🤔

File: harness/determined/core/_train.py
Comment (hamidzr) [reply harness/determined/core/_train.py:75]: > we allow anything that's json-serializable in metrics.

don't we disallow some non-numerical values in some cases? I recall getting bug reports about it.

While we are truly Dict[str, Any] it sounds good to expose it that way as well.

File: harness/determined/common/experimental/trial.py
Comment (hamidzr) [reply harness/determined/common/experimental/trial.py:374]: thanks for the comments! not forcing the user to look up what `MetricType` makes sense especially if they not using an ide that shows it right there eg in a notebook.

File: harness/determined/common/util.py
Comment (hamidzr) [reply harness/determined/common/util.py:36]: I don't know. these are not super clear to me. I can move them if you think they belong there.

File: harness/determined/common/util.py
Comment (hamidzr) [reply harness/determined/common/util.py:234]: I merged the two

File: harness/determined/common/experimental/determined.py
Comment (hamidzr) [reply harness/determined/common/experimental/determined.py:383]: we're no longer logging new deprecation warnings after https://github.com/determined-ai/determined/pull/7117/commits/286efe25d30525a25012d1f2a20587ab8f971086

File: harness/determined/common/experimental/trial.py
Comment (hamidzr) [reply harness/determined/common/experimental/trial.py]: I did this.

**Comment by hamidzr** [comment]
I un-deprecated the old methods so they won't be triggering any warning. took out type aliases. and avoided rebasing to keep the new changes easily visible.

File: harness/determined/core/_train.py
Comment (hamidzr) [inline harness/determined/core/_train.py:106]: what do you think should happen here? this was a non-existent branch before generic metrics.

Comment (hamidzr) [reply harness/determined/core/_train.py:106]: add a new method for tsb?
https://github.com/determined-ai/determined/blob/e23b58becdfe03a52fca01f031dce13e1287b340/harness/determined/tensorboard/metric_writers/callback.py#L21

Comment (hamidzr) [reply harness/determined/core/_train.py]: leaving it for a different branch/pr https://github.com/hamidzr/determined/tree/tsb-generic-metrics

File: harness/determined/core/_train.py
Comment (azhou-determined) [reply harness/determined/core/_train.py:72]: is the plan to change this to "series"? if so, should we wait for backend API name change? since if we land this, then change it later, it could become a user-facing change.

File: harness/determined/core/_train.py
Comment (azhou-determined) [inline harness/determined/core/_train.py:84]: is this check comment still relevant?

Comment (hamidzr) [reply harness/determined/core/_train.py:84]: we should chat about this.
prior to this pr we skipped these for training metrics do you know if that was by design or an oversight? 
```
        serializable_metrics = self._get_serializable_metrics(metrics)
        reportable_metrics = {k: metrics[k] for k in serializable_metrics}
```
do we see us not wanting to skip unserializable metrics before serializing to send metrics over the wire? 🤔

File: harness/determined/core/_train.py
Comment (hamidzr) [reply harness/determined/core/_train.py:72]: slack conversation webui,

**Comment by hamidzr** [comment]
it turns out we've been using metric_type in our public apis since 2020. I've made some changes to allow moving forward with the pr without exposing any metric_type publicly. 
https://github.com/determined-ai/determined/pull/7117/commits/983fa5a53305dc92dcf490f407966313685ef1bb
wdy think?

File: harness/determined/core/_train.py
Comment (tayritenour) [inline harness/determined/core/_train.py:70]: What if a user wants to report trial metrics to be able to save generic metrics? Shouldn't this be a public function so that one could make generic metrics? Or perhaps we should have a separate function for that?

Comment (hamidzr) [reply harness/determined/core/_train.py:70]: I agree the goal of doing this is to allow this to land before 
https://github.com/determined-ai/determined/pull/7117#issuecomment-1629652354 is resolved which is waiting for review https://github.com/determined-ai/determined/pull/7353 so that it's usable sooner.

once all is done we'd want to make these public but I leave that up to the mlsys team.

Comment (tayritenour) [reply harness/determined/core/_train.py:70]: Quick question, #7117 is this PR, did you mean a different one? Or perhaps I'm misreading, it seems like you were saying it was supposed to be a PR we are also waiting on in addition to #7353

File: harness/determined/core/_train.py
Comment (tayritenour) [reply harness/determined/core/_train.py:72]: Personally, I don't see any issue with `metric_type` for what it's worth. The PR to make this `group` is here: https://github.com/determined-ai/determined/pull/7353/files but honestly I would rather just let it be. Could we look into that PR @azhou-determined @rb-determined-ai ?

File: harness/determined/core/_train.py
Comment (tayritenour) [reply harness/determined/core/_train.py:84]: So if I'm understanding correctly, you are saying in the old metrics system we would skip metrics that don't serialize correctly? I don't know about the full conversation that led to that decision, but I would think this would be more confusing for the user to send certain invalid metrics and have them be silently ignored than to throw some kind of error/warning to say "this metric is invalid, please fix it"

File: harness/determined/core/_train.py
Comment (hamidzr) [reply harness/determined/core/_train.py:70]: > Quick question, https://github.com/determined-ai/determined/pull/7117 is this PR, did you mean a different one?
sry I could have been more clear. 

by 
> goal of doing this is to allow this to land before https://github.com/determined-ai/determined/pull/7117#issuecomment-1629652354 is resolved

I meant to say: before the issue of  'lets not proceed because maybe we want to expose a different name instead of `metric type` to end users ' is resolved

File: harness/determined/core/_train.py
Comment (hamidzr) [reply harness/determined/core/_train.py:84]: it looks like this is the case in `main` for some validation metrics but not training metrics (and I don't why the split happens) 
https://github.com/determined-ai/determined/blob/2ac22f00861bfc7c5700c47af5bde0f3f99949e5/harness/determined/core/_train.py#L142-L150

I have some guesses but let's take this conversation to #mlsys

File: harness/determined/common/experimental/determined.py
Comment (hamidzr) [inline harness/determined/common/experimental/determined.py:400]: take these out if we leave the generic version underscored.

File: harness/determined/common/experimental/trial.py
Comment (azhou-determined) [inline harness/determined/common/experimental/trial.py:375]: why is this wrapper method needed? why not just call `_stream_trials_metrics` directly?

Comment (hamidzr) [reply harness/determined/common/experimental/trial.py:375]: just following existing patterns to avoid more change than necessary. see stream_training_metrics

**Review by azhou-determined — APPROVED** [review]
lgtm

**Comment by hamidzr** [comment]
I'm getting "Forked prs not allowed to run on OSS projects" I'll move this to an upstream branch

**Comment by hamidzr** [comment]
thanks for the reviews.
migrated the pr over to https://github.com/determined-ai/determined/pull/7407
could you please stamp that as well?

**Pull request closed without merging**
